File: 01-Programming-Languages/R/Algorithms/Array/two_sum.R
# Problem: Two Sum
# Description: Find two indices in an array whose values sum to a target.
#
# Approach:
# - Use a hash map to store seen values and their indices
# - For each element, check if complement (target - element) exists
#
# Time Complexity: O(n)
# Space Complexity: O(n)
#
# Example:
# Input: nums = c(2, 7, 11, 15), target = 9
# Output: c(1, 2)

two_sum <- function(nums, target) {
  seen <- list()
  for (i in seq_along(nums)) {
    complement <- target - nums[i]
    if (!is.null(seen[[as.character(complement)]])) {
      return(c(seen[[as.character(complement)]], i))
    }
    seen[[as.character(nums[i])]] <- i
  }
  NULL
}

nums <- c(2, 7, 11, 15)
target <- 9
result <- two_sum(nums, target)
cat("nums:", nums, "\n")
cat("target:", target, "-> indices:", result[1], result[2], "\n")

File: 01-Programming-Languages/Scala/Algorithms/Array/TwoSum.scala
// Problem: Two Sum
// Description: Find two indices in an array whose values sum to a target.
//
// Approach:
// - Use a hash map to store seen values and their indices
// - For each element, check if complement (target - element) exists
//
// Time Complexity: O(n)
// Space Complexity: O(n)
//
// Example:
// Input: nums = Array(2, 7, 11, 15), target = 9
// Output: Array(0, 1)

object TwoSum {
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    val seen = scala.collection.mutable.Map[Int, Int]()
    for (i <- nums.indices) {
      val complement = target - nums(i)
      if (seen.contains(complement)) {
        return Array(seen(complement), i)
      }
      seen(nums(i)) = i
    }
    Array(-1, -1)
  }

  def main(args: Array[String]): Unit = {
    val nums = Array(2, 7, 11, 15)
    val target = 9
    val result = twoSum(nums, target)
    println(s"nums: ${nums.mkString(", ")}")
    println(s"target: $target -> indices: ${result(0)}, ${result(1)}")
  }
}
